src/lib.rs

```diff
@@ -132,45 +132,44 @@
 // In mDNS and DNS, the basic data structure is "Resource Record" (RR), where
 // in Service Discovery, the basic data structure is "Service Info". One Service Info
 // corresponds to a set of DNS Resource Records.
+use std::any::Any;
+use std::borrow::Borrow;
+use std::collections::{HashMap, HashSet};
+use std::convert::TryInto;
+use std::net::AddrParseError;
+use std::os::unix::io::RawFd;
+use std::str::FromStr;
+use std::time::SystemTime;
+use std::{cmp, fmt, str, thread, vec};
+
 use flume::{bounded, Sender, TrySendError};
 use log::{debug, error};
-use nix::{
-    errno, fcntl,
-    sys::{
-        select::{select, FdSet},
-        socket::{
-            bind, recvfrom, sendto, setsockopt, socket, sockopt, AddressFamily, InetAddr, IpAddr,
-            IpMembershipRequest, Ipv4Addr, MsgFlags, SockAddr, SockFlag, SockType,
-        },
-        time::{TimeVal, TimeValLike},
-    },
-};
-use std::{
-    any::Any,
-    cmp,
-    collections::{HashMap, HashSet},
-    convert::TryInto,
-    fmt,
-    os::unix::io::RawFd,
-    str, thread,
-    time::SystemTime,
-    vec,
+use nix::sys::select::{select, FdSet};
+use nix::sys::socket::{
+    bind, recvfrom, sendto, setsockopt, socket, sockopt, AddressFamily, InetAddr, IpAddr,
+    IpMembershipRequest, Ipv4Addr, MsgFlags, SockAddr, SockFlag, SockType,
 };
+use nix::sys::time::{TimeVal, TimeValLike};
+use nix::{errno, fcntl};
 
 /// A basic error type from this library.
-#[derive(Debug)]
+#[derive(Debug, PartialEq)]
 pub enum Error {
     /// Like a classic EAGAIN. The receiver should retry.
     Again,
 
     /// A generic error message.
     Msg(String),
+
+    /// Error during parsing of ip address
+    ParseIpAddr(String),
```

Comment:
are you sure? parsing the ip addr from string can result in an error, right now that error is just ignored, which I think could be surprising for new users.

Reply:
I think this will be useful if we change `ServiceInfo::new()` to return a `Result` (You mentioned in another place as well).

I'm wondering if we should take this opportunity to change `new()`, which is a breaking change of course.

Reply:
I'd be all for changing `new()`, even if it is a breaking change. I imagine new users might feel very frustrated and/or irritated if we just silence IP parsing errors.

Reply:
prob. offtopic here, but have you also thought about some reorganization of the project. Getting into the project was quite hard, as there's just a single 2.5k `lib.rs`. I think it would be beneficial for maintainability (and new features?) if we would split up the project into separate files. What do you think?